---
title: "JASEN results"
author: "Region Östergötland"
date: "`r Sys.Date()`"
link-citations: true
bibliography: bibliography.bib
biblio-style: "alpha" #https://www.overleaf.com/learn/latex/Bibtex_bibliography_styles
#csl: "pnas.csl"
subject: "This is a report over JASEN results"
params:
    sample: "[ID info missing]"
    multiqc: "[Link missing]"
keywords: 
  - JASEN
  - "Region Östergötland"
output:
  bookdown::html_document2:
    highlight: tango
    theme: yeti
    split_by: none # only generate a single output page
    self_contained: TRUE
    toc: yes
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      print: FALSE
    number_sections: FALSE
    code_download: FALSE
    pandoc_args: ["--top-level-division=section",
              "-V", "documentclass=report"]
  bookdown::pdf_document2:
    keep_tex: no
    #latex_engine: xelatex
    #highlight: pygments #Other options: default, tango, kate, monochrome, espresso, zenburn, haddock, and textmate
    #theme: spacelab
    # https://pandoc.org/MANUAL.html
    pandoc_args: ["--top-level-division=section",
                  "-V", "documentclass=article",
                  "-V", "linkcolor=MidnightBlue",
                  "-V", "citecolor=Aquamarine",
                  "-V", "urlcolor=NavyBlue",
                  "-V", "toccolor=NavyBlue",
                  "-V", "pagestyle=headings"]
---

**Sample ID: `r params$sample`**

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r include=FALSE}
library(jsonlite)
library(tidyverse)
library(DT)
library(bookdown)

dt_alise <- function(df, text_before ,html_object) {
  df %>%
  datatable( 
    rownames = FALSE,
    extensions = c('FixedColumns','Buttons','KeyTable'),
    options=list(
      scrollX=TRUE,
      dom = 'tBlrpf',
      lengthMenu = c(10, 30, 100),
      fixedColumns = list(leftColumns = 1),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
      keys = TRUE,
      pageLength = 30
      ), caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;',
    text_before , html_object)
    )
}

filter_ariba <- function(df){
  df %>% 
    filter(pc_ident > 99) %>%
    relocate(X.ariba_ref_name, .after = var_description) %>%
    relocate(pc_ident, .after = ref_name) %>%
    mutate(phenotype = str_replace(free_text, ".+\\|\\|\\|", "")) %>%
    relocate(phenotype, .after = pc_ident) %>%
    select(-free_text)
}

handle_empties <- function(df1, df2){
  # Both dfs are empty, return the first empty one only
  if ((dim(df1)[1] == 0 & dim(df1)[2] == 0) & (dim(df2)[1] == 0 & dim(df2)[2] == 0)) {
    return(df1)
  # df1 empty, df2 has something, return df2
  }else if((dim(df1)[1] == 0 & dim(df1)[2] == 0) & (dim(df2)[1] != 0 & dim(df2)[2] != 0)){
    return(df2)
  # df2 empty, df1 has something, return df1  
  }else if((dim(df1)[1] != 0 & dim(df1)[2] != 0) & (dim(df2)[1] == 0 & dim(df2)[2] == 0)){
    return(df1)    
  }else if((dim(df1)[1] != 0 & dim(df1)[2] != 0) & (dim(df2)[1] != 0 & dim(df2)[2] != 0)){
    # Make data types to be equal between df:s
    common <- names(k4d)[names(k4d) %in% names(k5d)]
    k4d[common] <- lapply(common, function(x) {
      match.fun(paste0("as.", class(k5d[[x]])))(k4d[[x]])
    })
    return(bind_rows(k4d,k5d))
  }else{
    return(data.frame(c("Something","wrong"),c("went","!")))
  }
}


k1 <- fromJSON("cgmlst_stats.json", flatten=TRUE)
k2 <- fromJSON("cgmlst_alleles.json", flatten=TRUE)
k3 <- fromJSON("mlst.json", flatten=TRUE)
k4 <- fromJSON("motif_report_local.json", flatten=TRUE)
k5 <- fromJSON("motif_report_nonc.json", flatten=TRUE)
k6 <- fromJSON("motif_report.json", flatten=TRUE)
k7 <- fromJSON("quast_report.json", flatten=TRUE)
k8 <- fromJSON("snp_report.json", flatten=TRUE)

k1d <- do.call(rbind.data.frame, k1)
k2d <- do.call(rbind.data.frame, k2)
k4d <- do.call(rbind.data.frame, k4)
k5d <- do.call(rbind.data.frame, k5)
k6d <- do.call(rbind.data.frame, k6)
k7d <- do.call(rbind.data.frame, k7)
k8d <- do.call(rbind.data.frame, k8)
phenos <- read_tsv("phenotypes.tsv")

#options(DT.options = list(, buttons = c('copy', 'csv', 'excel', 'pdf', 'print')))
```


# Cgmlst Stats

```{r echo=FALSE}
# Remove default column names and ...
oldnames <- do.call(rbind.data.frame, k1) %>%
  colnames()
# ... use these instead:
newnames <- c("EXC","INF","LNF","PLOT","NIPH","ALM","ASM")
df <- do.call(rbind.data.frame, k1) %>%
  rename_at(vars(all_of(oldnames)), ~ newnames)
df <- df[-1, ]

dt_alise(df, "Cgmlst stats")
```

# Cgmlst Alleles

```{r echo=FALSE, warning=FALSE}
# Move first row to be the column names
names(k2d) <- NULL
names(k2d) <- lapply(k2d[1, ], as.character)
k2d <- k2d[-1, ]

# Transpose df
k2d <- as_tibble(cbind(names(k2d), t(k2d)), 
                 validate = FALSE) 
names(k2d) <- c("gene_name", "allele")

k2d %>%
  datatable(
    rownames = FALSE,
    extensions = c('FixedColumns','Buttons','KeyTable'),
    options=list(
      scrollX=TRUE,
      dom = 'tBlrpf',
      lengthMenu = c(10, 30, 100, 500),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
      keys = TRUE,
      pageLength = 10
      ), caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;',
    'Table: ', htmltools::em("Cgmlst Alleles"))
    )

```

# Mlst

```{r echo=FALSE}
k3 %>%
  datatable(
    rownames = FALSE,
    extensions = c('FixedColumns','Buttons','KeyTable'),
    options=list(
      scrollX=TRUE,
      dom = 'tBlrpf',
      lengthMenu = c(10, 30, 100, 500),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
      keys = TRUE,
      pageLength = 10
      ), caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;',
    'Table: ', htmltools::em("Mlst"))
    )
```

Reference: [@Silva2018]

# Ariba 

## Local database motif report (coding and non-coding)

```{r echo=FALSE}
handle_empties(k4d,k5d) %>%
  filter_ariba %>%
  dt_alise('Table 4: Ariba run output for local antibiotics resistance genes. For detailed information on the table headers, see: ', htmltools::tags$a(href="https://github.com/sanger-pathogens/ariba/wiki/Task:-run#report-file", "ariba run subcommand output description"))
```


## Resfinder motif report

```{r echo=FALSE, warning=FALSE}
# If something was found from Resfinder
if (dim(k6d)[1] != 0 & dim(k6d)[2] != 0) {
  # Make join names homogeneous accross the two tables
  phenos$join_column = str_replace_all(phenos[["Gene_accession no."]],"\\(|\\)|\\'|-","_")
  k6d$join_column = str_replace_all(k6d$ref_name,"\\.","_")
  
  # Join phenotype information and print as a pretty table
  left_join(k6d,phenos, by=c("join_column" = "join_column")) %>% 
    filter(pc_ident > 99) %>% 
    relocate(X.ariba_ref_name, .after = var_description) %>%
    relocate(pc_ident, .after = ref_name) %>%
    dt_alise('Table 5: Ariba run output for resistance genes from Resfinder. For detailed information on the table headers, see: ', htmltools::tags$a(href="https://github.com/sanger-pathogens/ariba/wiki/Task:-run#report-file", "ariba run subcommand output description"))
}else{
  # If not found, return empty
  k6d %>% 
    dt_alise('Table 5: Ariba run output for resistance genes from Resfinder. For detailed information on the table headers, see: ', htmltools::tags$a(href="https://github.com/sanger-pathogens/ariba/wiki/Task:-run#report-file", "ariba run subcommand output description"))
}
```

Reference: [@Zankari2012]

## Summary

```{r echo=FALSE, message=FALSE, warning=FALSE}
read_csv("summary.csv") %>%
  datatable( 
    rownames = FALSE,
    extensions = c('FixedColumns','Buttons','KeyTable'),
    options=list(
      scrollX=TRUE,
      dom = 'tBlrpf',
      lengthMenu = c(5, 10, 25),
      #fixedColumns = list(leftColumns = 1),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
      keys = TRUE,
      pageLength = 10
      ), caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;',
    'Table: ', htmltools::em("Summary"))
    )
```

# Quast report

```{r echo=FALSE}
t(k7d) %>%
  datatable(
    rownames = FALSE,
    extensions = c('FixedColumns','Buttons','KeyTable'),
    options=list(
      scrollX=TRUE,
      dom = 'tBlrpf',
      lengthMenu = c(10, 30, 100),
      #fixedColumns = list(leftColumns = 1),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
      keys = TRUE,
      pageLength = 30
      ), caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;',
    'Table 6: Output of genome assembly quality assessment tool. For detailed information on the table headers, see: ', htmltools::tags$a(href="http://quast.sourceforge.net/docs/manual.html#sec3.1.1", "QUAST manual"))
    )
```

# SNP report

```{r echo=FALSE}
    dt_alise('Table 7: Ariba run output for resistance genes from Resfinder. For detailed information on the table headers, see: ', htmltools::tags$a(href="https://github.com/sanger-pathogens/ariba/wiki/Task:-run#report-file", "ariba run subcommand output description"))
```

# MultiQC report

MultiQC report can be found [here](`r params$multiqc`). Reference to MultiQC: @Ewels2016.

# References
